Codes/Clean Data/DataProperties_20200318.R

```diff
@@ -0,0 +1,115 @@
+rm(filename)
+###GET TIMESTAMPS FROM MESSAGE_1 FILE
+filename<-paste0(directory1,"/",folders1[i],"/",firms[i],"_",dates[j],"_","34200000_57600000_message_1.csv")
+data_m<-read.table(file=filename,sep=",",stringsAsFactors=F,col.names=c("time","eventtype","orderID","size","price","direction","mpid"))
+rm(filename)
+###MERGE THEM
+data_o$time<-data_m$time
+data_o$clocktime<-format(as.POSIXct((data_o$time),origin="1970-01-01",tz="UTC"),"%H:%M:%OS6")
+data_o<-unique(data_o)
+rm(data_m)
+data_o<-as.data.table(data_o)
+setkey(data_o,clocktime)
+###LOAD MESSAGE_0 FILE
+filename<-paste0(directory2,"/",folders2[i],"/",firms[i],"_",dates[j],"_","14400000_72000000_message_0.csv")
+data_m<-read.table(file=filename,sep=",",stringsAsFactors=F,col.names=c("time","eventtype","orderID","size","price","direction","mpid"))
+rm(filename)
+data_m$clocktime<-format(as.POSIXct((data_m$time),origin="1970-01-01",tz="UTC"),"%H:%M:%OS6")
+data_m<-as.data.table(data_m)
+setkey(data_m,clocktime)
+###MERGE DATASETS
+data<-data_o[data_m,roll=T,mult="last"]
+data$i.time<-NULL
+rm(data_o,data_m)
+
+###STANDARD CLEANING ALGORITHM
+#only keep messages that are part of the continuous trading period
+#data<-data[data$time>=startTrad&data$time<=endTrad,]
+#check for trading halts
+tradehaltIdx<-which(data$eventtype==7&data$direction==-1);
+tradequoteIdx<-which(data$eventtype==7 & data$direction==0);
+traderesumeIdx<-which(data$eventtype==7 & data$direction==1);
+  if(length(tradehaltIdx)==0 & length(tradequoteIdx)==0  & length(traderesumeIdx)==0 ){print("No trading halts detected.")}
+  if(length(tradehaltIdx) !=0){cat("Data contains trading halt! at time stamp(s)", data$clocktime[tradehaltIdx],"\n" )}
+  if(length(tradequoteIdx) !=0){cat(" Data contains quoting message! at time stamp(s)", data$clocktime[tradequoteIdx], "\n")}
+  if(length(traderesumeIdx) !=0){cat(" Data resumes trading! at time stamp(s) ", data$clocktime[traderesumeIdx],"\n")}
+rm(tradehaltIdx,tradequoteIdx,traderesumeIdx)
+
+mpidratio<-sum(data$mpid!="null")/nrow(data)
+mpidnum<-length(unique(data$mpid[data$mpid!="null"]))
+mpidmean<-mean(table(data$mpid[data$mpid!="null"]),na.rm=T)
+mpidmedian<-median(table(data$mpid[data$mpid!="null"]),na.rm=T)
+mpidsd<-sd(table(data$mpid[data$mpid!="null"]),na.rm=T)
+mpidskew<-skewness(table(data$mpid[data$mpid!="null"]),na.rm=T)
+mpidhhi<-sum(((table(data$mpid[data$mpid!="null"])/sum(table(data$mpid[data$mpid!="null"])))*100)^2,na.rm=T)
+
+fullstats[j,i,1]<-mpidratio
+fullstats[j,i,2]<-mpidnum
+fullstats[j,i,3]<-mpidmean
+fullstats[j,i,4]<-mpidmedian
+fullstats[j,i,5]<-mpidsd
+fullstats[j,i,6]<-mpidskew
+fullstats[j,i,7]<-mpidhhi
+
+f<-paste0(directory3,"/",firms[i])
+if (file.exists(f)==F){dir.create(f)}
+filename<-paste0(f,"/",firms[i],"_",dates[j],".Rds")
+save(data,file=filename)
+
+
+}
+}
+
+filename<-paste0("W:/LOBSTER/MPIDStats_v1_20200323.Rds")
+save(fullstats,file=filename)
+
+filename<-paste0("W:/LOBSTER/MPIDStats_v1_20200323.Rds")

```

Comment:
It may be better to put the stats part code to a separated script, so that the user will not run the heavy pre-process code every time.

Reply:
I will create two different codes. "Data Properties" will contain summary statistics calculations, and "Clean Merge Data" will run the first part of the code, which merges orderbook and message files together.